pkg/cover/manager_to_ci.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package cover
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+// CIDetails fields will be added to every CSV line.
+type CIDetails struct {
+	Timestamp      string `json:"timestamp"`
+	FuzzingMinutes int    `json:"fuzzing_minutes"`
+	Arch           string `json:"arch"`
+	Manager        string `json:"manager"`
+	KernelRepo     string `json:"kernel_repo"`
+	KernelBranch   string `json:"kernel_branch"`
+	KernelCommit   string `json:"kernel_commit"`
+	Compiler       string `json:"compiler"`
+}
+
+type dbCoverageRecort struct {
+	CIDetails
+	FilePath     string `json:"file_path"`
+	FunctionName string `json:"function_name"`
+	TotalCount   int    `json:"total_cb_count"`
+	CoveredCount int    `json:"covered_cb_count"`
+}
+
+func serializeJSLine(w io.Writer, covInfo *dbCoverageRecort) error {
+	bs, err := json.Marshal(covInfo)
+	if err != nil {
+		return fmt.Errorf("failed to marshal covInfo: %w", err)
+	}
+	bs = append(bs, []byte("\n")...)
+	if _, err = w.Write(bs); err != nil {
+		return fmt.Errorf("failed to write js data: %w", err)
+	}
+	return nil
+}
+
+// ManagerToCI generates GCP template pipeline file.
+// Every line of this file will be converted into 1 DB row.
+// Every line is an independent JSON struct.
+// For the fuzzing session level records, file name and function name are empty.
+// For the file level records, function name is empty.
+func ManagerToCI(jsonBytes []byte, ciDetails CIDetails) ([]byte, error) {
+	covInfo := new(coverageInfo)
+	if err := json.Unmarshal(jsonBytes, covInfo); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal coverage: %w", err)
+	}
+	var result bytes.Buffer
+	w := bufio.NewWriter(&result)
```

Comment:
You don't need bufio.Writer to write to bytes.Buffer.
bytes.Buffer is not an os.File/net.Conn and is fast on its own.
Using bufio in this case only adds additional memory allocations and copies.

Reply:
Thanks. Changed to w := new(bytes.Buffer).